### ISSUE TITLE
How to know requirements for inputs and outputs of mkldnn_primitive_create for various primitives

### ISSUE THREAD

**Issue by blueberry** [intel/mkl-dnn#307]
I'm refering to the C API, but it probably also holds to the C++ API. Various desc's and primitive_desc's are more or less defined by the API, meaning there is (more or less) enough information in the API itself to know what arguments are required, in which order, etc.

However, when it comes to the `mkldnn_primitive`, it is one generic "catch-all" function that creates primitives for everything, from memory, via inference forward operations and training forward operations, to bacward operations. The trouble is that each of these has many variations depending on the actual operation, actual alg_kind, etc. It is enough to introduce a slight change in the parameters of a descriptor, to implicitly change the requirements for primitive's creation function. 

The question is simple: How to know the requirements that implicitly go into `mkldnn_primitive_create` and similar functions? How do I know that when I am creating a primitive for the operation X with alg_kind Y and variation Z, I have to provide a workspace memory as the third element in a dst array for example?

**Comment by emfomenk** [comment]
Hi @blueberry,

Yeah, our C-API is not well documented with this respect. Sorry for inconvenience.
Currently you have 3 options to understand what is the right order of the inputs/outputs:

1. look at C++ API -- it calls C-API underneath, so you can extract the required information from there. Example [here](https://github.com/intel/mkl-dnn/blob/v0.16/include/mkldnn.hpp#L2958). Note how many constructors we have.

2. look into the actual implementation, better reference. E.g. for batch normalization (the primitive with the biggest number of flavors) see [here](https://github.com/intel/mkl-dnn/blob/v0.16/src/cpu/ref_batch_normalization.cpp#L31). You can check there that batch normalization expects source to be the first input, while mean can either be the second input (if stats are provided by user) or the second output (if mean and variance are to be calculated by the primitive).

Workspace (if required) is typically passed as the last output. To understand whether you need workspace or not you should [query](https://github.com/intel/mkl-dnn/blob/v0.16/include/mkldnn_types.h#L1122) the workspace_pd from the primitive descriptor. If it is empty then ws is not needed. Otherwise you need to create and pass it to the creation function.

I don't like this particular part about our C-API. To be honest I don't like creation part in C++-API even more -- it requires you write separate code for different configuration, which leads to boilerplate code like [this](https://github.com/intel/mkl-dnn/blob/v0.16/tests/gtests/test_batch_normalization.cpp#L391) and [this](https://github.com/intel/caffe/blob/1.1.1a/src/caffe/layers/mkldnn_batch_norm_layer.cpp#L341) -- 40 lines, srsly???..

To address this issue we are going to change the creation and execution API soon. There would be two main changes:
1. creation would not require any memory, it would just create a primitive. The inputs and outputs would be passed at execution

2. instead of vector of input/output memories themselves we would ask to pass a `map<type_tag, memory>`, e.g. batch-norm run would look like:
```cpp
  mkldnn_memory_t src, scaleshift; // inputs
  mkldnn_memory_t dst, mean, var, ws; // outputs

  // the next line is just to give an idea
  typedef struct { mkldnn_tag_t tag, mkldnn_memory_t mem } *mkldnn_exec_resources_t;

  // the order of the entries can be arbitrary
  mkldnn_exec_resources_t res = { 
    {mkldnn_tag_src, src}, {mkldnn_tag_dst, dst},
    {mkldnn_tag_mean, mean}, {mkldnn_tag_variance, var},
    {mkldnn_tag_scale_shift, scaleshift},
    {mkldnn_tag_workspace, ws}
  };
  mkldnn_primitive_exec(bnorm, res);
```

Just out of the curiosity why do you use C-API and not C++-API?

**Comment by blueberry** [comment]
I am creating a Clojure high-level API that uses Java bindings underneath, and it is much more straightforward to me to use C api despite its verbosity because it enables me to have precise control of memory allocations and deallocations.

**Comment by emfomenk** [comment]
oK, I see. Actually C++-API allows you to do the same. You can always pass your pointers to the memory, so memory management should be possible with that API as well.

WRT documentation, I will put the notes to the corresponding init_descriptor, so that there would be at least some documentation what order and what memories the primitives are expecting.

Thanks for your feedback!

**Comment by blueberry** [comment]
In my experience, it is very difficult to control due to the Java<->C++ mismatch, especially since JavaCPP bindings have their own memory deallocators that try to rely on Java's GC lifecycle management.  Works on simple examples, completely breaks down under stress. With the C API, I can do my own memory managent, not only of memory, but of all primitives, and it works rather well (with the usual caveat that it requires meticulous care to clean up everything right).

**Comment by emfomenk** [comment]
I see. Thanks!

**Comment by blueberry** [comment]
Thank you for working on this!
